rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+  matched by the pattern and predicate. Unlike the `PatternRewriter`, candidates are free to overlap. Mainly used
+  to bring BYOC patterns into the Collage framework.
+- `OpPredicatePartitionRule`: Given an attribute name, produces a candidate for every call to a primitive Relay
+  operator where the operator i) has predicate bound to that attribute which ii) returns true given the
+  call sub-expression. Generally this will result in a singleton sub-graph containing only the call, but it may also
+  pull in constant arguments to the call should they be required. Used to bring BYOC operator predicates into the
+  Collage framework.
+- `OpCallByKindPartitionRule`: Uses the `"TOpPattern"` attribute provided for every Relay operator to produce a
+  candidate for every call to a 'fusable Relay operator'. Used to select the operators which `FuseOps` will consider
+  parts of kernels.
+
+The combinator rules are:
+
+- `CompositePartitionRule`: Indicates all sub-candidates matched by the sub-rule should be wrapped by a `"Composite"`
+  function. The `"Composite"` name is taken from the rule name. Used to indicate Relay operators (or groups of Relay
+  operators) should be mapped to target-specific operators, both for BYOC and TVM external library integrations.
+- `PrimitivePartitionRule`: Indicates all sub-candidates matched by the sub-rule should be wrapped by a `"Primitive"`
+  function, possibly with an additional `"Compiler"` attribute. Used to delineate a partition (or kernel).
+- `UnionPartitionRule`: Simply unions all the sub-candidates from all sub-rules together. Used to combine
+  individual `DFPatternPartitionRules`.
+- `CombinePartitionRule`: Given a sub-rule and a list of 'combiner' rules (see below), finds all possible ways of
+  combining the sub-candidates to yield even larger candidates. Note that the sub-candidates may also be directly
+  included in the results. The 'combiner' rules allow combining by `OpPatternKinds`, combining the arguments to
+  tuples which themselves are arguments to Relay operator calls, and so on. This rule is intended to mimic the
+  existing TVM `FuseOps` pass, though: i) all candidates are found rather than just the largest, ii) the starting
+  set of candidates can be provided by any other rule, and iii) we rely on `SubGraph` validity checking to weed out
+  infeasible candidates.
+- `OnlyValidPartitionRule`: Given a `SubGraphConfig`, ignores candidates with 'invalid' sub-graphs. Used to limit the
+  maximum candidate depth, the number of independent outputs, and whether intermediate 'taps' are allowed.
+- `HostPartitionRule`: Produces candidates for all Relay expressions which could be
+  'left behind' for execution by the host (eg on the VM). This rule lets us simplify the overall Collage search
+  algorithm.
+
+Here are some typical ways to combine `PartitionRules` for different partition flavors:
```

Comment:
How are these rules populated ?

Can they be "registered" to avoid unnecessary search ?
The registration could be allowed or a dis-allowed list.

Reply:
Added a forward reference to the 'Phase 1' section. Yeah, I'd like to allow the Target to directly specify the rule to use by a "partition_rule" attribute, but since that's not a simple atomic scalar I'm not sure how the ergonomics will work out. Otherwise there's logic to, effectively, 'import' the existing BYOC patterns etc by inspecting the global pattern table (and UMA could help make this a little more structured I think).